### ISSUE TITLE
Toast accessibility issue

### ISSUE THREAD

**Issue by Shemetiuk** [bootstrap-vue/bootstrap-vue#3322]
### Describe the bug

The toast is read multiple times.

### Steps to reproduce the bug

You can use buttons on toast page
https://bootstrap-vue.js.org/docs/components/toast/
On Mac with VoiceOver turned on, it reads twice on Safari, 3 times on Chrome. 
On Windows FF/NVDA, it only reads when there are two toasts.

### Expected behavior

It should read once

### Versions

**Libraries:**

- BootstrapVue: 2.0.0-rc.20

### Demo link

https://bootstrap-vue.js.org/docs/components/toast/

**Comment by tmorehouse** [comment]
We are using Bootstrap's aria markup based on their documentation examples: https://getbootstrap.com/docs/4.3/components/toasts/#placement

```html
<!-- Outer Toaster wrapper -->
<div aria-live="polite" aria-atomic="true" ...>
  <!-- intermediate positioning wrapper -->
  <div ...>
    <div class="toast" role="alert" aria-live="assertive" aria-atomic="true">
      <div class="toast-header">...</div>
      <div class="toast-body">...</div>
    </div>
    <div class="toast" role="alert" aria-live="assertive" aria-atomic="true">
      <div class="toast-header">...</div>
      <div class="toast-body">...</div>
    </div>
  </div>
</div>
```

Perhaps it is the double nested `aria-live` regions that is causing the duplicate announcements.

**Comment by Shemetiuk** [comment]
You're right. It's the first thing which I did. But even after deleting one of the aria-live, Chrome with Voice Over reads it twice and Jaws or FF/NVDA read only if open two toasts.

**Comment by tmorehouse** [comment]
Just doing a bit of googling, noticed this comment at WAI-ARIA:

> role=alert which is equivalent to using aria-live=assertive

So I wonder if providing both `role="alert"` and `aria-live="assertive"` on the `.toast` element may be the culprit?

**Comment by Shemetiuk** [comment]
Good point. What I've found is that different browsers with different screen readers can have different behavior for role and aria-live that's why probably people often use both at the same time. Unfortunately I don't have laptop with me right now. But I'll definitely try on Monday play with it and let you know.

**Comment by tmorehouse** [comment]
Or maybe the following should be used (when toasts are in a toaster):

```html
<!-- Outer Toaster wrapper -->
<div aria-live="polite" aria-atomic="false" ...>
  <!-- intermediate positioning wrapper -->
  <div ...>
    <!-- individual toasts -->
    <div class="toast" role="alert" aria-atomic="true">
      <div class="toast-header">...</div>
      <div class="toast-body">...</div>
    </div>
    <div class="toast" role="alert" aria-atomic="true">
      <div class="toast-header">...</div>
      <div class="toast-body">...</div>
    </div>
  </div>
</div>
```

**Comment by tmorehouse** [comment]
One other thing that may help is delaying the insertion of the toast's header/body once the toast is added to the document (as aria-live regions should be in dom before the content is inserted. i.e. by maybe move the aria-live/aria-atomic to the b-toast wrapper div that wraps each toast (not shown in the above code) as the toast div is actually delayed because it is the transition element of the toast.

**Comment by tmorehouse** [comment]
I've installed ChromeVox on chrome, and can hear the duplicate announcement.  Even when removing the aria-* attributes off of the toaster.

I think this is either related to to transition on the toast/toaster, or a weird interaction with Portal-Vue.

The "static" toast (not rendered in a b-toaster) reads only once.

I'm going to try with no aria-* attributes on the toast, but have the aria-* on the b-toaster, and see if hte toasts are still read twice.

**Comment by tmorehouse** [comment]
It appear that the exit transition is causing the issues.

I think we can address this by dynamically removing the aria attributes on the toast component before the toast closes.

**Comment by Shemetiuk** [comment]
I already tried to remove all aria attributes for toaster and live it only for toast. And tried opposite too. Chrome read twice in both cases

**Comment by tmorehouse** [comment]
Yeah, same thing I experienced.  I think dynamically removing the `aria-*` off the toast before they start to hide might fix this.  and removing the aria-attributes completely off of the toaster component.

**Comment by tmorehouse** [comment]
It is definitely the leave transition that is triggering the duplicate read.  I set the auto-hide-delay to 20 seconds and the toast is re-read on exit.  almost as if the aria-live region is speaking additions and removals

**Comment by tmorehouse** [comment]
@Shemetiuk check out the deploy preview https://deploy-preview-3329--bootstrap-vue.netlify.com/docs/components/toast for the possible fix.It is working for me now using ChromeVox on windows

**Comment by Shemetiuk** [comment]
@tmorehouse thank you for your fast responses. I tested this page and here's results:
**VoiceOver** - Chrome and Safari (both read once so they're fine)
**Jaws** - Chrome and Firefox is fine, but IE doesn't read single toast (you should open two and then it will read the latest one)
**NVDA** - Chrome and Firefox is fine, IE doesn't read.

**Safari:**
VoiceOver should works with Safari because it's Mac soft so I can say it works properly with this fix. Just reminder: In the current version v2.0.0-rc.20 - it reads twice.
**Firefox:**
NVDA is the best pair with Firefox and it works fine with this fix. Just reminder: In the current version v2.0.0-rc.20 - you should open two toasts and then it will read the latest one.
**IE:**
Jaws should works properly with IE, but unfortunately it's not working with this fix (you should open two toasts and then it will read the latest one). Just reminder: In the current version v2.0.0-rc.20 - it reads one toast twice, when toast is opened and when it closed.

**Comment by tmorehouse** [comment]
With the latest proposed fix, the toast content is appended one animation frame after the toast wrapper is rendered (the toast wrapper has the `aria-live`/`aria-atomic` region defined).

So there must be some issue with how IE sees the live region being updated... maybe one animation frame isn't enough of a delay for it.  I know the live regions must exist in the DOM before any content is updated in the region.

So maybe the toaster should have `aria-live="polite"`, `aria-atomic="false"` and `aria-relevant="additions"`, while the toasts have `aria-live="true" and `aria-atomic="true"` (but have the `aria-*` removed just before the toast is transitioned out.

I'm wondering if the reading would be affected wether if the toast is `appended` or `prepended` ?

**Comment by Shemetiuk** [comment]
Toast by itself should has ability to receive prop is-status (as it works now) to change role and aria-live. Because in different cases you need different roles/behaviors.

To have in the beginning aria-* for toaster and toast can be not a good idea because it's how it works now, and it reads twice with VoiceOver etc. So I feel that after this fix we have less issues but IE still doesn't work properly.

I don't think that using aria-relevant can be helpful for us in some case. And I don't think that append or prepend could make some difference in how it will reads by screen readers.

Do you think changing animation frame can works for IE?

**Comment by tmorehouse** [comment]
Maybe adding in one more $nextTick() to delay the toast content showing might help...

**Comment by tmorehouse** [comment]
@Shemetiuk I have added an extra `$nextTick()` delay to the showing of the toast content in the aria-live region.   If you get a chance, could you try the deploy preview again with IE?

**Comment by tmorehouse** [comment]
I've downloaded NVDA, and tried in IE 11, and the toasts are not spoke (unless I add aria-live to the toaster component), however using NVDA with Chrome, the toasts are spoken (once and a while spoken twice, but mostly spoken once).

ChromeVox works fine for Chome.

I found an issue at teh NVDA rep that was closed in 2014, but some of the discussion sounds familiar to the double speak issue https://github.com/nvaccess/nvda/issues/4066 (i.e. spoken twice in some browsers and sometimes never in IE 11).

**Comment by Shemetiuk** [comment]
@tmorehouse 
VoiceOver with Safari - works
VoiceOver with Chrome - doesn't work

Jaws with IE - doesn't work

NVDA with Firefox - works.

So yesterday version worked better.

**Comment by tmorehouse** [comment]
So for IE 11 and NVDA, the only way I can get toasts to be announced (when using $bvToast.toast()) was to have the following on the b-toaster outer DIV:

```html
aria-live="assertive" aria-atomic="false" aria-relevant="additions text"
```

And have the standard `role="alert" aria-live="assertive" aria-atomic="true"` on the toast (as is today).

worse come to worst, we may need to browser sniff for IE 11, and set those values on the toaster (if not overridden by the user)

**Comment by tmorehouse** [comment]
I was not quite right with my previous comment.  IE worked if I manually edited the toaster attributes (in the inspector), but when adding it to the toaster code, it did not work.

After a bit more googling...

From https://terrillthompson.com/tests/aria/live-scores.html
> NVDA in IE11 reads the live region twice each time it's updated. It does not do this in Firefox. I tried changing the Game 3 announcement to **role="alert" but then NVDA does read the announcement at all (in IE11)**.
> ...
> Conclusions: In order to get the updated scores to be read properly by VoiceOver in Mac OS X, NVDA in IE, and Window-Eyes in Firefox, the method used in Game 3 is necessary (all updated content in one container, not in separate containers within a live region parent). The container for the updated content should be marked up with aria-live rather than role="alert", since the latter does not seem to be supported at all using NVDA in IE. However, even the Game 3 method is not accessible in JAWS 16 in IE (the solution then is on users: Don't use JAWS 16 in IE).

It sounds like there might be a conflict with having both `role="alert"` and `aria-live="..."` with IE 11 and NVDA

**Comment by tmorehouse** [comment]
So I think we are SOL when it comes to IE11 support for screen readers and toasts. :(

But at least the toasts are now read only once for other browsers.

**Comment by Shemetiuk** [comment]
@tmorehouse Thank you for your help. Everything works except of IE11 and it's pretty good. I will think about it and if I'll find some solution for IE11 I'll let you know.

**Comment by tmorehouse** [comment]
@Shemetiuk yep, the changes will be in 2.0.0-rc.21

**Comment by Shemetiuk** [comment]
@tmorehouse just want to mention that accessibility works good after these changes. Thank you
